File: WOT-iOS/WOTPivot/WOTPivot/core/adapters/webRequest/WOTRequestExecutor.m
//
//  WOTRequestExecutor.m
//  WOT-iOS
//
//  Created on 6/2/15.
//  Copyright (c) 2015. All rights reserved.
//

#import "WOTRequestExecutor.h"
#import "WOTLogger.h"
#import <WOTPivot/WOTPivot.h>
#import <WOTPivot/WOTPivot-Swift.h>

@interface WOTRequestExecutor()

@property (nonatomic, strong) NSMutableDictionary *grouppedRequests;
@property (nonatomic, readwrite, assign) NSInteger pendingRequestsCount;

@end

@implementation WOTRequestExecutor

+ (WOTRequestExecutor *)sharedInstance {
    
    static WOTRequestExecutor* instance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        
        instance = [[WOTRequestExecutor alloc] init];
    });
    
    return instance;
}

+ (NSString *)pendingRequestNotificationName {
    return @"WOTRequestExecutorPendingRequestNotificationName";
}

- (id)init {
    
    self = [super init];
    if (self){
        
        self.pendingRequestsCount = 0;
        self.grouppedRequests = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (void)setPendingRequestsCount:(NSInteger)pendingRequestsCount {
    _pendingRequestsCount = pendingRequestsCount;
    NSString *notificationName = [WOTRequestExecutor pendingRequestNotificationName];
    [[NSNotificationCenter defaultCenter] postNotificationName: notificationName object:self];
}

- (void)cancelRequestsByGroupId:(NSString *)groupId {
    
    NSPointerArray *requests = self.grouppedRequests[groupId];
    if ([requests count] == 0){

        return;
    }
    
    [requests compact];
    for (NSInteger idx = ([requests count]-1); idx>=0; idx--) {
        
        WOTRequest *request = [requests pointerAtIndex:idx];
        [request cancel];
    }
    
    NSCAssert(([requests count] == 0), @"requestsArray is not empty after cancelation");
}

- (void)requestsArray:(NSPointerArray *)requests removeRequest:(id<WOTRequestProtocol>)request {
    
    [requests compact];
    
    NSCAssert([requests count] != 0, @"requestsArray is empty");
    
    NSInteger requestIndex = [[requests allObjects] indexOfObject:request];
    if (requestIndex != NSNotFound) {
        [requests removePointerAtIndex:requestIndex];
    } else {
        debugError(@"attempting to remove unknown request:%@",request);
    }
}

- (BOOL)add:(id<WOTRequestProtocol>)request byGroupId:(NSString *)groupId {
    
    NSPointerArray *requests = self.grouppedRequests[groupId];
    [requests compact];
    
    if (!requests){
        
        requests = [[NSPointerArray alloc] init];
        self.grouppedRequests[groupId] = requests;
    }
    
    NSUInteger index = [[requests allObjects] indexOfObjectPassingTest:^BOOL(WOTRequest *testingRequest, NSUInteger idx, BOOL *stop) {
        
        BOOL isEqual = [testingRequest isEqual:request];
        if (isEqual) {
            
            *stop = YES;
        }
        return isEqual;
    }];
    
    BOOL canAdd = (index == NSNotFound);
    if (canAdd) {
        [request addListener: self];
        [request addGroup:groupId];
        [requests addPointer:(__bridge void *)request];
    }
    return canAdd;
}

- (id<WOTRequestProtocol>)createRequestForId:(NSInteger)requestId {

    id<WOTRequestProtocol> result =  [[WOTRequestReception sharedInstance] createRequestForRequestId:requestId];
    result.hostConfiguration = self.hostConfiguration;
    return result;
}

#pragma mark - WOTRequestListener

- (void)request:(id)request finishedLoadData:(NSData *)data error:(NSError *)error {

    Class clazz = [request class];
    
    if ([clazz conformsToProtocol:@protocol(WOTModelServiceProtocol)]) {
        id<WOTModelServiceProtocol> modelService = (id<WOTModelServiceProtocol>)clazz;
        
        NSString *modelClassName = [modelService modelClassName];
        NSArray* requestIds = [[WOTRequestReception sharedInstance] requestIdsForClass: NSClassFromString(modelClassName)];

        for (NSNumber* requestId in requestIds) {
            [[WOTRequestReception sharedInstance] requestId:[requestId integerValue] processBinary:data error:error];
        }
    }

    self.pendingRequestsCount -= 1;
    [self removeRequest:request];
}

- (void)requestHasCanceled:(id<WOTRequestProtocol>)request {
    
    self.pendingRequestsCount -= 1;
    [self removeRequest:request];
}

- (void)requestHasStarted:(id<WOTRequestProtocol>)request {
    self.pendingRequestsCount += 1;
}

- (void)removeRequest:(id<WOTRequestProtocol>)request {
    
    NSArray *groups = [request availableInGroups];
    [groups enumerateObjectsUsingBlock:^(id group, NSUInteger idx, BOOL *stop) {
        
        NSPointerArray *requests = self.grouppedRequests[group];
        [self requestsArray:requests removeRequest:request];
    }];
    [request removeListener:self];
}

@end

File: WOT-iOS/WOTPivot/WOTPivot/core/adapters/webRequest/WOTRequestExecutor.swift
//
//  WOTRequestExecutor.swift
//  WOTPivot
//
//  Created by Pavel Yeshchyk on 3/10/20.
//  Copyright © 2020 Pavel Yeshchyk. All rights reserved.
//

import Foundation

@objc(WOTRequestExecutorSwift)
public class WOTRequestExecutorSwift: NSObject, WOTRequestManagerProtocol {
    
    @objc
    required public  init(requestReception: WOTRequestReceptionProtocol) {
        reception = requestReception
        super.init()
    }
    
    @objc
    public static let pendingRequestNotificationName: String = "WOTRequestExecutorPendingRequestNotificationName"
    
    fileprivate var grouppedRequests: [String: [WOTRequestProtocol]] = [:]
    
    @objc
    public var reception: WOTRequestReceptionProtocol

// WOTRequestExecutorSwift

    @objc
    public var pendingRequestsCount: Int = 0

    @objc
    public func addRequest(_ request: WOTRequestProtocol, forGroupId groupId: String) -> Bool {
        var grouppedRequests: [WOTRequestProtocol] = []
        if let available = self.grouppedRequests[groupId] {
            grouppedRequests.append(contentsOf: available)
        }
        let filtered = grouppedRequests.filter { (availableRequest) -> Bool in
            availableRequest.hash == request.hash
        }
        let result: Bool = (filtered.count == 0)
        if result {
            request.addListener(self)
            request.addGroup(groupId)
            grouppedRequests.append(request)
        }
        self.grouppedRequests[groupId] = grouppedRequests
        return result
    }
    
    @objc
    public func cancelRequests(groupId: String) {
        guard let requests = self.grouppedRequests[groupId], requests.count > 0 else {
            return
        }
        requests.forEach { request in
            request.cancel()
        }
    }
}


extension WOTRequestExecutorSwift: WOTRequestListenerProtocol {
    
    public var requestReception: WOTRequestReceptionProtocol {
        get { return reception }
        set { reception = newValue }
    }

    public func request(_ request: AnyObject, finishedLoadData data: Data?, error: Error?, invokedBy: WOTNestedRequestsEvaluatorProtocol) {
        if let clazz = type(of: request) as? NSObject.Type {
            if let instance = clazz.init() as? WOTModelServiceProtocol {
                if let modelClass = instance.instanceModelClass() {
                    let requestIds = requestReception.requestIds(forClass: modelClass)
                    requestIds?.forEach({ requestId in
                        requestReception.requestId(requestId, processBinary: data, error: error, invokedBy: invokedBy)
                    })
                }
            }
        }
        pendingRequestsCount -= 1
        #warning ("force cast")
        self.removeRequest(request as! WOTRequestProtocol)
    }
    
    public func requestHasCanceled(_ request: WOTRequestProtocol) {
        pendingRequestsCount -= 1
        removeRequest(request)
    }
    
    public func requestHasStarted(_ request: WOTRequestProtocol) {
        pendingRequestsCount += 1
    }
    
    public func removeRequest(_ request: WOTRequestProtocol) {
        request.availableInGroups.forEach { group in
            if var grouppedRequests = self.grouppedRequests[group] {
                grouppedRequests.removeAll(where: { $0.hash == request.hash })
                self.grouppedRequests[group] = grouppedRequests
            }
            
        }
        request.removeListener(self)
    }
}
